### ISSUE TITLE
 Ignoring the body of 304 replies causes corrupted messages

### ISSUE THREAD

**Issue by sonertari** [e2guardian/e2guardian#136]
Within a persistent connection:
- If the reply from the HTTP server (OpenBSD/httpd) is 304, e2guardian/dansguardian assumes that there is no body, hence sets its content length to 0. This decision leaves the message body waiting in the buffer.
- When the next reply arrives, e2guardian/dansguardian fetches everything in the buffer including the previous reply's body, which of course is at the beginning.
- Obviously, this confuses e2guardian/dansguardian, because it now thinks that the previous reply's body is the current reply's header, hence sends a corrupted header to the client.
- This causes the web browser (Firefox) to keep waiting until the persistent connection times out, which might take a long time.

There are HTTP servers which send a body with 304 messages, such as the new OpenBSD/httpd. Therefore, ignoring the body of 304 replies is causing corrupted messages, hence should be avoided.

I am attaching a patch which fixes this issue and two evidence files, a sample 304 reply from OpenBSD/httpd and a debug output of e2guardian. Pay special attention to the logs where e2guardian changes header.front() to "<!DOCTYPE /".

[patch-src_HTTPHeader_cpp.txt](https://github.com/e2guardian/e2guardian/files/423513/patch-src_HTTPHeader_cpp.txt)
[OpenBSD_httpd_304.txt](https://github.com/e2guardian/e2guardian/files/423511/OpenBSD_httpd_304.txt)
[debug.txt](https://github.com/e2guardian/e2guardian/files/423512/debug.txt)

**Comment by fredbcode** [comment]
Thanks yes very interesting, I guess this situation is a performance issue 
I will push your patch to the dev branch

**Comment by fredbcode** [comment]
ok pushed to https://github.com/e2guardian/e2guardian/tree/develop

**Comment by sonertari** [comment]
Great, thanks

**Comment by fredbcode** [comment]
Reopen
Need more tests in production before closing 
Thanks for your help

**Comment by fredbcode** [comment]
Hi sonertari

Sorry but there is a performance issue with your patch 
I'm trying the develop's branch with or without your patch 
eg: http://www.nextinpact.com/ -> 10s with patch-src_HTTPHeader_cpp.txt 2s without

**Comment by fredbcode** [comment]
What do you think of something like ?

 108     // code 304 - not modified - no content
 109     String temp(header.front().after(" "));
 110     if (temp.startsWith("304") && (pcontentlength == NULL)){
 111         contentlength = 0;
 112     }   
 113     else if (pcontentlength != NULL) {
 114         String temp(header.front().after(" "));
 115         temp = pcontentlength->after(":"); 
 116         contentlength = temp.toOffset();
 117     }

**Comment by fredbcode** [comment]
Hum, maybe something like this should be better (sorry I have no time for testing right now)

    String temp(header.front().after(" "));

    // most usual case body is not empty
    if (pcontentlength != NULL) {
        temp = pcontentlength->after(":");
        contentlength = temp.toOffset();
    } else if (temp.startsWith("304") && (pcontentlength == NULL)){
        contentlength = 0;
    }

**Comment by sonertari** [comment]
Hi Frederic,

As I have explained in my bug report, this issue causes corrupted messages, hence it is a bug. So, if I have to choose between a performance penalty and a bug, I would of course prefer the performance penalty.

On the other hand, I think that any performance decrease should not be a concern here, because we should filter any/all content: (1) if a 304 response has a body, we should scan not discard it, (2) if not, then how can a zero length (non-existent) body cause a performance issue?

Nonetheless, if we still want to ignore any 304 body, we should definitely fetch any 304 body from the buffer, along with its headers, so that the buffer is ready for the next message.

So the following additional patch fetches and discards any 304 body. But I do not recommend it, because it seems like yet another hack to me (ConnectionHandler.cpp seems quite complicated as it is, any plans of refactoring it?). Beware though that the dm_plugin may still wait for the proxy (?):

[patch-src_ConnectionHandler_cpp.txt](https://github.com/e2guardian/e2guardian/files/435926/patch-src_ConnectionHandler_cpp.txt)

**Comment by fredbcode** [comment]
I agree about your point, but the performance issue is really huge with many websites.
I don't know why but with or without your fix there is a real difference, not for you ?

With many users the response time is very low

Thanks for your help, I will take a look soon

**Comment by fredbcode** [comment]
> Nonetheless, if we still want to ignore any 304 body, we should definitely fetch any 304 body

Hum, I certainly missed something sorry 

I thought in my last code below that I take care of it ? 
304 && pcontentlength == NULL is removed but 304 with body = contentlength 

Sent from my ASUS

**Comment by philipianpearce** [comment]
I do not agree that we should process or scan a body sent with a 304. 

Sending a body with a 304 is a protocol violation, so there is a bug in OpenBSD/httpd which needs fixing, as it is highly likely to cause problems with other web clients (browsers etc). 

As far as the e2g code, my view is that we should apply Postel's Law. i.e. accept and ignore the body sent in error, making sure we send the browser a response which complies with the HTML standards. There is no way we should process or send the body, as this could be used as a method of attack on client software. 

Agree ConnectionHandler.cpp is over complicated and difficult to read. I have added a lot of comments (beginning/ending marks) to the legacy code so that it is a bit more readable and I can now find my way around better than before, but it is still not well structured as I would like! It is something that will be looked at in the next major version. 

Regards 

Philip 

----- Original Message -----

**Comment by sonertari** [comment]
Hi Philip,

Yes, rfc2616 p.33 reads ".. 304 (not modified) responses MUST NOT include a message-body." And yes, this should be considered as a bug in OpenBSD/httpd, and it must be fixed. So, alright, I have changed my mind, we don't have to scan the 304 body.

But, still, ignoring the 304 body corrupts the header of the next message. And, this is a bug in e2guardian, and it must be fixed. (e2guardian assigns 0 to the contentlength var only, it does not even call setContentLength(0) to fix the 304 header to client. Not to mention leaving the 304 body in the buffer.)

So, e2guardian should fetch and *discard* not ignore the 304 body. It should also fix the contentlength field of the 304 header. See the two patches I have provided, when combined they achieve exactly those.

Btw, I have seen the changes and comments in ConnectionHandler.cpp (the last time I checked the Dansguardian code was in 2006). Keep up the good work guys.
